### PR TITLE
Add support for net451 projects.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Previous to this any net451 projects that Razor-tooling would attempt to load would explode.
- We now dispatch to ourselves within a users output to ensure that all dependencies are matched up correctly.
- Added framework/configuration options to enable dispatching to properly find where our counterpart exists in a projects output.
- Renamed bin directory option to build base. With dispatching logic VisualStudio can't easily determine the users RID which is required for an output path to dotnet.
- Added a new hidden argument `--no-dispatch` to indicate when we're calling our counterpart in a projects output so we don't recurse forever.
- Replaced `AssemblyLoadContext` logic with `Assembly.Load` since we now always invoke ourselves in an environment where `Assembly.Load` is accurate. This way logic in the desktop process and non-desktop processes are identical.
- Worked around an issue in `ProjectDependenciesCommandFactory` where configuration isn't properly flowed from constructor to `Create` method: https://github.com/dotnet/cli/issues/2512
- Worked around an issue where binding redirects were not being created for executable desktop packages. This involved upping Newtonsoft.Json to 8.0.3: dotnet/cli#2505

#55

Playing around with what it'd mean to have a full end-to-end test for this guy. That'll be a separate PR/issue since it'll be pretty extensive.

/cc @anurse @natemcmaster @prafullbhosale

### CODE REVIEW

File: src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs [hunk 1]
```diff
@@ -4,92 +4,236 @@
...
                 config.HelpOption("-?|-h|--help");
                 var protocolOption = config.Option(
                     "-p|--protocol",
                     "Protocol to resolve TagHelperDescriptors with.",
                     CommandOptionType.SingleValue);
-                var binDirectory = config.Option(
-                    "-b|--bin-dir",
-                    "The project's bin directory.",
+                var buildBasePathOption = config.Option(
+                    "-b|--build-base",
```
Comment (NTaylorMullen) [inline src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs]: Long name should be `--build-base-path` to stay consistent with `dotnet`

**Comment by NTaylorMullen** [comment]
Put this into release, not dev.

File: src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs [hunk 2]
```diff
@@ -4,92 +4,236 @@
...
+                .GetTargetFrameworks()
+                .Select(frameworkInformation => frameworkInformation.FrameworkName);
 
-            if (projectContext == null)
+            NuGetFramework framework;
+            if (!TryResolveFramework(frameworkOption, targetFrameworks, projectFilePath, out framework))
             {
-                throw new InvalidOperationException(
-                    string.Format(CultureInfo.CurrentCulture, Resources.InvalidProjectFile, projectFilePath));
+                // Could not resolve framework for dispatch. Error was reported, exit early.
```
Comment (natemcmaster) [inline src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs]: Log this?

File: src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs [hunk 3]
```diff
@@ -4,92 +4,236 @@
...
+                return 0;
             }
 
-            return projectContext;
+            var configurationValue = configurationOption.Value() ?? Constants.DefaultConfiguration;
+            var projectContext = new ProjectContextBuilder()
```
Comment (natemcmaster) [inline src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs]: Have you tested standalone apps yet? I'm just curious if we need to also specify runtimes here.

File: src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs [hunk 4]
```diff
@@ -4,92 +4,236 @@
...
+                .WithProject(projectFile)
+                .WithTargetFramework(framework)
+                .Build();
+            var buildBasePathValue = buildBasePathOption.Value();
+            var commandFactory = new ProjectDependenciesCommandFactory(
+                projectContext.TargetFramework,
+                configurationValue,
+                outputPath: null,
```
Comment (natemcmaster) [inline src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs]: Why did you leave out outputPath but include build base path?

File: src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs [hunk 5]
```diff
@@ -4,92 +4,236 @@
...
+            {
+                dispatchArgs.Add("--protocol");
+                dispatchArgs.Add(protocolOption.Value());
+            }
+
+            using (var errorWriter = new StringWriter())
+            {
+                var commandExitCode = commandFactory
+                    .Create(
+                        PlatformServices.Default.Application.ApplicationName,
```
Comment (natemcmaster) [inline src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs]: May need to change, dependency on the outcome of https://github.com/dotnet/cli/issues/2592

File: src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs [hunk 6]
```diff
@@ -4,92 +4,236 @@
...
+                }
+
+                return 0;
+            }
+        }
+
+        private static int Run(CommandArgument assemblyNamesArgument, CommandOption protocolOption)
+        {
+            var protocol = protocolOption.HasValue() ?
+                int.Parse(protocolOption.Value()) :
```
Comment (natemcmaster) [inline src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs]: Woudl recomment `int.TryParse` to be safe

File: src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs (hunk 4 above)
Comment (NTaylorMullen) [inline src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs]: Today, razor tooling is only built to be consumed by Visual Studio. They only pass the build base path.

File: src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs (hunk 5 above)
Comment (NTaylorMullen) [inline src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs]: :+1:

File: src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs]: Will log to verbose

File: src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs (hunk 3 above)
Comment (NTaylorMullen) [inline src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs]: Gonna give this a shot. Good idea.

Comment (NTaylorMullen) [inline src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs]: Talked to Andrew offline and utilizing the runtimes current candidate rids should suffice.

File: src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/dotnet-razor-tooling/Internal/ResolveTagHelpersCommand.cs]: Actually, this is logged in a different format within ResolveFramework, going to leave as is.

**Comment by NTaylorMullen** [comment]
:up: :date:

**Comment by NTaylorMullen** [comment]
Going to close this and merge with the PR that consumes dotnettooldispatcher

**Pull request closed without merging**